README
What is this?

    A Python API and script for the retrieval and storage of Play Whe results
    from the National Lotteries Control Board (NLCB) [http://www.nlcb.co.tt/].

Why was this written?

    Short answer:
        Because it can be written.

    Slightly longer answer:
        Because I was looking for a small Python project to work on for practice.

Which version do I have?

    Version 0.4.

    After installing the package you can also check the version number by issuing
    the following command at a shell prompt.

    $ playwhe.py --version

What has changed from version 0.3 to 0.4?

    On November 21st, 2011 the NLCB started drawing Play Whe 3 times per day.
    Due to this change the display format of the results retrieved via their
    website was changed. Since this program simply scrapes the HTML off their
    results page, I needed to update the script to work with the new format.

    Hence, the logic for retrieving and storing the Play Whe results have
    changed. However, the interface to perform these operations have not.

How do I install it?

    $ tar xvzf playwhe-0.4.tar.gz
    $ cd playwhe-0.4
    $ sudo python setup.py install

What are some of the things it can do?

    You can get Play Whe results directly from NLCB using the playwhe module.

    > import playwhe
    > p = playwhe.PlayWhe()

    > # retrieve and display all the results for the month of April in the year 2011
    > print "\n\n".join(map(lambda r: r.prettyprint(), p.results_for_month(2011, 4)))

    > # retrieve and display the results for April 2nd, 2011
    > print "\n\n".join(map(lambda r: r.prettyprint(), p.results_for_day(2011, 4, 2)))

    > # retrieve and display the three most recent results
    > print "\n\n".join(map(lambda r: r.prettyprint(), p.results()))

    You can get Play Whe results directly from NLCB using the playwhe.py script.

    $ echo Retrieve and display the results for the month of April in the year 2011
    $ playwhe.py --pretty-print --date=2011-04
    $ playwhe.py -p -d 2011-04

    $ echo Retrieve and display the results for April 2nd, 2011
    $ playwhe.py --pretty-print --date=2011-04-02
    $ playwhe.py -p -d 2011-04-02

    $ echo Retrieve and display the three most recent results
    $ playwhe.py --pretty-print
    $ playwhe.py -p

    You can keep a local copy of previous Play Whe results in an SQLite database.

    $ echo Create and initialize a Play Whe database
    $ playwhe.py --createdb="/home/<username>/playwhe.db"
    $ playwhe.py -c "/home/<username>/playwhe.db"

    $ echo Update the Play Whe database
    $ playwhe.py --updatedb="/home/<username>/playwhe.db"
    $ playwhe.py -u "/home/<username>/playwhe.db"

    $ echo Update the Play Whe database and keep a log
    $ playwhe.py -u "/home/<username>/playwhe.db" 2>> "/home/<username>/playwhe.log"

Where can I get the latest development version of this project?

    This project is hosted on GitHub at https://github.com/dwayne/playwhe.

    $ git clone git://github.com/dwayne/playwhe.git

What liscense does this project use?

    This project is in the public domain. Do with it whatever you want.

Where can I get help?

    > import playwhe
    > help(playwhe)

    $ playwhe.py --help

    You can also get help, report bugs, make suggestions or ask thoughtful
    questions by contacting Dwayne R. Crooks via email at <me at dwaynecrooks dot com>.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.8.0-alpha.2 (2019-03-16)

A refactoring of the library to have it expose a more useful public API.

### Added

- The `playwhe.client` module
- The `playwhe.cli` module
- A notes file, `NOTES.md`
- An environment variable condition on tests that use the real server. The
  environment variable is called `PLAYWHE_TESTS_USE_REAL_SERVER`

### Removed

- The `playwhe.scraper` module
- The `playwhe.db` module
- The `bin/results.sh` script

### Changed

- The public API

## 0.8.0-alpha.1 (2018-10-20)

A complete rewrite.

### Added

- The `playwhe.scraper` module
- The `playwhe.db` module
- The `bin/results.sh` script for saving results from a Play Whe database to a
  CSV file
- An archive of Play Whe results that's stored in the `data/results.csv` CSV
  file
- Support for logging
- A suite of tests
- A license, MIT
- A changelog

### Changed

- The CLI

## 0.7.1 (2017-12-10)

I forgot to make some important updates to the results in the Play Whe database
after making the changes in 0.7.

## 0.7 (2016-04-01)

The NLCB website occasionally has incorrect results. More often than not the
draw numbers are the culprit. However, if we order the data by date, then by
period and then by draw number we can mitigate any ill effects this bad data
causes.

### Removed

- Backup databases since it wasn't necessary to keep them in this project

### Fixed

- How data is retrieved so that invalid draw numbers don't affect the
  correctness of the data being displayed

## 0.6 (2015-07-28)

On July 6th, 2015 the NLCB started drawing Play Whe 4 times per day. Due to this
the `playwhe.py` script was updated to scrape the new 4:00pm draw.

### Changed

- `playwhe.py` to scrape the new 4:00pm draw
- Period from a number, one of `1, 2, 3`, to a string, one of `EM, AM, AN, PM`

### Fixed

- Bad results data in the backup Play Whe database

## 0.5 (2015-03-29)

Around February 21st, 2015 the NLCB took down their website to redesign and
rebuild it. They changed the service URL along with the format of the HTML
results. Due to this, the parser had to be updated.

### Added

- A few notes

### Changed

- The parser to parse results based on the new syntax
- The service URL

### Fixed

- The name of the Play Whe results backup file
- The regular expression used to parse results so that it ignores case

## 0.4 (2011-12-11)

On November 21st, 2011 the NLCB started drawing Play Whe 3 times per day. Due to
this, the display format of the results on their website changed. Since this
program simply scrapes the HTML off their results page, the script was updated
to work with the new format.

### Added

- Play Whe results up to December 10th, 2011

### Changed

- The `playwhe.py` script to work with NLCB's new format

### Fixed

- The long description in `setup.py`

## 0.3 (2011-06-28)

### Added

- A database to store Play Whe results
- Play Whe results from July 4th, 1994 to June 28th, 2011

### Changed

- The README
- The short description in `setup.py`

## 0.2 (2011-06-25)

### Added

- A README
- A `Mark` class

### Changed

- `playwhe.py` so that it can be used as both a script and module
- Rename `Draw` to `Result`
- The documentation
- The error messages

### Fixed

- `name_of_mark` bug
- `self.__mark` bug

## 0.1 (2011-05-14)

Initial release


playwhe.py
#! /usr/bin/env python

"""A Python API and script for retrieving and storing Play Whe results.

The library provides a Python interface for retrieving Play Whe results from
the National Lotteries Control Board (NLCB) website at http://www.nlcb.co.tt/.

The script uses the library to provide a tool for the retrieval and storage
of Play Whe results.

"""

import datetime
import re
import sys

from operator import attrgetter
from urllib import urlencode
from urllib2 import urlopen, URLError

__version__ = "0.7"
__author__  = "Dwayne R. Crooks"
__email__   = "me@dwaynecrooks.com"

# Play Whe's birthday
start_date = datetime.date(1994, 7, 4) # July 4th, 1994

# The date that Play Whe changed to having 3 draws per day
three_draws_date = datetime.date(2011, 11, 21) # November 21st, 2011

# The date that Play Whe changed to having 4 draws per day
four_draws_date = datetime.date(2015, 7, 6) # July 6th, 2015

# the abbreviated months of the year
month_abbr = ["",
    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
]

class Mark(object):
    """A Play Whe mark."""

    lowest  = 1  # the lowest Play Whe mark
    highest = 36 # the highest Play Whe mark

    # associate the marks with their standard name
    name_of_number = {
        1: "centipede",
        2: "old lady",
        3: "carriage",
        4: "dead man",
        5: "parson man",
        6: "belly",
        7: "hog",
        8: "tiger",
        9: "cattle",
       10: "monkey",
       11: "corbeau",
       12: "king",
       13: "crapaud",
       14: "money",
       15: "sick woman",
       16: "jamette",
       17: "pigeon",
       18: "water boat",
       19: "horse",
       20: "dog",
       21: "mouth",
       22: "rat",
       23: "house",
       24: "queen",
       25: "morocoy",
       26: "fowl",
       27: "little snake",
       28: "red fish",
       29: "opium man",
       30: "house cat",
       31: "parson wife",
       32: "shrimps",
       33: "spider",
       34: "blind man",
       35: "big snake",
       36: "donkey"
    }

    def __init__(self, number):
        if number < Mark.lowest or number > Mark.highest:
            raise ValueError("number is out of range")

        self.number = number
        self.name = Mark.name_of_number[number]

    @staticmethod
    def get_name_of_number(number):
        """Return the standard name associated with the given number."""
        if number < Mark.lowest or number > Mark.highest:
            raise ValueError("number is out of range")

        return Mark.name_of_number[number]

    def __repr__(self):
        return '%s(%d, "%s")' % (self.__class__, self.number, self.name)

    def __str__(self):
        return "%d (%s)" % (self.number, self.name)

class Result(object):
    """A Play Whe result."""

    def __init__(self, draw, date, period, number):
        """Create a result.

        draw        - a positive integer that uniquely identifies the result
        date        - a datetime.date object
        period      - 'EM', 'AM', 'AN' or 'PM', indicating whether the draw was
                      at 10:30 AM, 1:00 PM, 4:00 PM or 6:30 PM
        number      - a number between Mark.lowest and Mark.highest inclusive
        """
        if draw < 1:
            raise ValueError("draw is out of range")

        if not isinstance(date, datetime.date):
            raise TypeError("date is not a datetime.date object")

        if period not in ["EM", "AM", "AN", "PM"]:
            raise ValueError("period is invalid")

        if number < Mark.lowest or number > Mark.highest:
            raise ValueError("number is out of range")

        self.draw = draw
        self.date = date
        self.period = period
        self.number = number

    @property
    def numeric_period(self):
        """Return a numeric representation of the period that's useful for sorting."""
        return ["EM", "AM", "AN", "PM"].index(self.period)

    def __repr__(self):
        return '%s(%d, %s, %s, %d)' % \
            (self.__class__,
             self.draw,
             repr(self.date),
             self.period,
             self.number)

    def __str__(self):
        """Return a string representing the result in CSV format.

        For example, str(Result(1, playwhe.start_date, 'AM', 15)) == "1,1994-07-04,AM,15".
        """
        return "%d,%s,%s,%d" % \
            (self.draw,
             self.date.isoformat(),
             self.period,
             self.number)

    def prettyprint(self):
        """Return a nicely formatted string representing the result."""
        return "Date:   %s-%s-%d\nDraw:   %d\nPeriod: %s\nNumber: %d (%s)" % \
                (str(self.date.day).zfill(2),
                 month_abbr[self.date.month],
                 self.date.year,
                 self.draw,
                 self.period,
                 self.number,
                 Mark.get_name_of_number(self.number))

class PlayWheException(Exception):
    pass

class PlayWhe(object):

    def __init__(self,
                 host="nlcb.co.tt",
                 service="/app/index.php/pwresults/playwhemonthsum",
                 timeout=15):
        self.host = host
        self.service = service
        self.timeout = timeout

    def results_for_month(self, year, month):
        """Return a list of results for each day in the given month and year.

        The results are ordered in increasing order of the draw number.
        """
        if year == start_date.year and month == start_date.month:
            date = start_date
        else:
            date = datetime.date(year, month, 1)

        if date < start_date or date > datetime.date.today():
            return []

        params = urlencode({
            "year" : str(year % 100).zfill(2),
            "month": month_abbr[month]
        })

        try:
            f = urlopen("http://" + self.host + self.service,
                        params,
                        self.timeout)
            html = f.read()
            f.close()
        except IOError, AttributeError:
            raise PlayWheException("Unable to retrieve the results for %d-%s" % \
                (year, str(month).zfill(2)))

        return self._parse_results(year, month, html)

    def results_for_day(self, year, month, day):
        """Return a list of results for the given day.

        The results are ordered in increasing order of the draw number.
        """
        date = datetime.date(year, month, day) # performs date validation

        try:
            results = self.results_for_month(year, month)
        except PlayWheException:
            raise PlayWheException("Unable to retrieve the results for %d-%s-%s" % \
                (year, str(month).zfill(2), str(day).zfill(2)))

        return filter(lambda r: r.date == date, results)

    def results(self):
        """Return a list with the three most recent results.

        The results are ordered in increasing order of the draw number.
        """
        date = datetime.date.today()
        date = datetime.date(date.year, date.month, 1)

        try:
            results = self.results_for_month(date.year, date.month)
        except PlayWheException:
            raise PlayWheException("Unable to retrieve the results for the previous two drawings")

        while len(results) < 3:
            date = date - datetime.timedelta(1)
            date = datetime.date(date.year, date.month, 1)
            try:
                results = self.results_for_month(date.year, date.month) + results
            except PlayWheException:
                raise PlayWheException("Unable to retrieve the results for the previous two drawings")

        return results[-3:]

    def _parse_results(self, year, month, html):
        # html is a string of HTML containing Play Whe results of the form:
        #     <h2>
        #         <strong> Draw #: </strong>13721<br>
        #         <strong> Date: </strong>04-Mar-15<br>
        #         <strong> Mark Drawn: </strong>19<br>
        #         <strong> Drawn at: </strong>EM<br>
        #     </h2>
        pattern = r"Draw #: </strong>(\d+).*? Date: </strong>(\d{1,2})-%s-%s.*? Mark Drawn: </strong>(\d+).*? Drawn at: </strong>(EM|AM|AN|PM)" % \
            (month_abbr[month], str(year % 100).zfill(2))

        results = []
        for r in re.findall(pattern, html, re.IGNORECASE):
            try:
                result = Result(int(r[0]),
                                datetime.date(year, month, int(r[1])),
                                r[3].upper(),
                                int(r[2]))
                results.append(result)
            except (ValueError, TypeError), e:
                print >> sys.stderr, 'IntegrityError: day(%d) draw(%d) period("%s") number(%d)' % (int(r[1]), int(r[0]), r[3], int(r[2]))

        return sorted(results, key=attrgetter('date', 'numeric_period', 'draw'))

import sqlite3

def createdb(db_path):
    """Create and initialize a Play Whe database.

    db_path should be an absolute path to the database. For example,
    /home/<username>/playwhe.db.
    """
    try:
        conn = sqlite3.connect(db_path)
    except sqlite3.Error:
        raise PlayWheException("Sorry, unable to connect to the database at %s." % db_path)

    c = conn.cursor()

    # setup the initial tables and relationships
    c.execute("""
    CREATE TABLE IF NOT EXISTS marks(
        number INTEGER PRIMARY KEY,
        name TEXT
    )""")

    c.execute("""
    CREATE TABLE IF NOT EXISTS results(
        draw INTEGER PRIMARY KEY,
        date TEXT,
        period TEXT NOT NULL,
        number INTEGER NOT NULL REFERENCES marks(number)
    )""")

    # populate the marks table
    for number, name in Mark.name_of_number.iteritems():
        c.execute("INSERT OR IGNORE INTO marks VALUES(?,?)", (number, name))

    conn.commit()
    conn.close()

network_connection_error = \
    "Sorry, unable to retrieve the results at this time.\n" + \
    "Please check your network connection and try again at a later time."

def updatedb(db_path):
    """Update a Play Whe database with the latest results.

    db_path should be an absolute path to the database. For example,
    /home/<username>/playwhe.db.
    """
    try:
        conn = sqlite3.connect(db_path)
    except sqlite3.Error:
        raise PlayWheException("Sorry, unable to connect to the database at %s." % db_path)

    conn.row_factory = sqlite3.Row
    c = conn.cursor()

    try:
        c.execute("""
        SELECT * FROM RESULTS
        ORDER BY
            DATE DESC,
            CASE PERIOD
                WHEN 'EM' THEN 3
                WHEN 'AM' THEN 2
                WHEN 'AN' THEN 1
                WHEN 'PM' THEN 0
            END,
            DRAW DESC
        LIMIT 1
        """)
    except sqlite3.Error:
        raise PlayWheException("Sorry, problems encountered accessing the database at %s." % db_path)

    last_result = c.fetchone()

    last = (start_date, "AM")
    if last_result:
        date = datetime.date(*map(int, last_result["date"].split('-')))
        if last_result["period"] == "EM":
            last = (date, "AM")
        elif last_result["period"] == "AM":
            last = (date, "AN")
        elif last_result["period"] == "AN":
            last = (date, "PM")
        else:
            last = (date + datetime.timedelta(days=1), "EM")
    last_date, last_time_of_day = last
    current_date = datetime.date.today()

    try:
        p = PlayWhe()
        for year in range(last_date.year, current_date.year + 1):

            start_month = last_date.month if year == last_date.year else 1
            end_month = current_date.month if year == current_date.year else 12

            for month in range(start_month, end_month + 1):
                print >> sys.stderr, "[%s] Fetching results for %s-%s..." % (datetime.datetime.now().strftime("%Y-%m-%d %I:%M%p"), year, str(month).zfill(2)),

                try:
                    results = p.results_for_month(year, month)
                except PlayWheException:
                    raise PlayWheException(network_connection_error)
                for r in results:
                    conn.execute("INSERT OR IGNORE INTO results (draw, date, period, number) VALUES(?,?,?,?)", (r.draw, r.date.isoformat(), r.period, r.number))
                    conn.commit()

                print >> sys.stderr, "DONE!"
    finally:
        conn.close()

if __name__ == "__main__":
    from optparse import OptionParser
    import os.path

    parser = OptionParser(usage="usage: %prog [options]",
                          version=__version__,
                          description="A script for the retrieval and storage of Play Whe results.",
                          epilog="For more help or to report bugs, please contact %s at %s." % (__author__, __email__))
    parser.add_option("-c", "--createdb", dest="createdb_path", metavar="PATH",
                      help="create and initialize a Play Whe database. PATH must "
                           "be a path to the database you want to setup. For "
                           "example, /home/<username>/playwhe.db")
    parser.add_option("-u", "--updatedb", dest="updatedb_path", metavar="PATH",
                      help="update a Play Whe database with the latest results. "
                           "PATH must be a path to the database you want to "
                           "update. For example, /home/<username>/playwhe.db")
    parser.add_option("-d", "--date",
                      help="display all the results for DATE. DATE must take "
                           "one of the formats: yyyy-mm or yyyy-mm-dd. If DATE "
                           "has the format yyyy-mm then all the results for "
                           "the month in that year are displayed. Otherwise, "
                           "DATE has the format yyyy-mm-dd, and all the "
                           "results for that day are displayed")
    parser.add_option("-p", "--prettyprint",
                      action="store_true", default=False,
                      help="display results in a nice human readable format")

    (options, args) = parser.parse_args()

    try:
        def display_results(results):
            if results:
                if options.prettyprint:
                    print "\n\n".join(map(lambda r: r.prettyprint(), reversed(results)))
                else:
                    print "\n".join(map(str, reversed(results)))
            else:
                print "No results found."

        if options.createdb_path:
            createdb(os.path.abspath(options.createdb_path))
        elif options.updatedb_path:
            print "Press Ctrl-C to safely exit at anytime."
            print

            updatedb(os.path.abspath(options.updatedb_path))
        elif options.date:
            def results_for(year, month, day=None):
                if day is not None:
                    return PlayWhe().results_for_day(year, month, day)
                return PlayWhe().results_for_month(year, month)

            try:
                display_results(results_for(*map(int, options.date.split("-"))))
            except ValueError, TypeError:
                parser.error('Must be a valid date in the format "yyyy-mm" or "yyyy-mm-dd"')
            except PlayWheException:
                print network_connection_error
                sys.exit(1)
        else:
            try:
                display_results(PlayWhe().results())
            except PlayWheException:
                print network_connection_error
                sys.exit(1)
    except PlayWheException, e:
        print e
        sys.exit(1)
    except KeyboardInterrupt:
        print
    except Exception, e:
        print "Sorry, an unknown error has occurred. Program terminated abnormally."
        sys.exit(1)
